### ISSUE TITLE
Unsound `Send`/`Sync` bounds

### ISSUE THREAD

**Issue by steffahn** [KaiJewson/try-rwlock#3]
```rs
// No bounds on `T` are required because the write guard's existence ensures exclusive access (so
// `T` becomes irrelevant).
unsafe impl<T, U: Send> Send for WriteGuard<'_, T, U> {}
unsafe impl<T, U: Sync> Sync for WriteGuard<'_, T, U> {}
```
I don’t even understand what the comment on this `impl` is supposed to tell me, but it’s unsound anyways :-)

Repro:
```rs
use std::cell::Cell;
use try_rwlock::{TryRwLock, WriteGuard};

// run in debug mode for clearly visible UB
fn main() {
    let mut unit = ();
    let cell = Cell::new(Ok::<&[usize], &[()]>(&[]));
    let lock = TryRwLock::new(&cell);
    // this guard is `Send`!
    let guard: WriteGuard<'_, &Cell<_>, ()> = WriteGuard::map(lock.try_write().unwrap(), |_| &mut unit);
    let bad_slice = std::thread::scope(|s| {
        let handle = s.spawn(|| {
            // now we got the `&Cell` into another thread!
            let r: &Cell<_> = *WriteGuard::unmap(guard);
            // racey-race…
            loop {
                let s = r.get();
                if let Ok(v) = s {
                    if !v.is_empty() {
                        break v;
                    }
                }
            }
        });
        let r = &cell;
        // racey-race…
        loop {
            r.set(Ok(&[]));
            r.set(Err(&[(); usize::MAX]));
            if handle.is_finished() {
                break;
            }
        }
        handle.join().unwrap()
    });
    println!("Here’s all your memory :-)\n{bad_slice:?}");
}
```

**Comment by steffahn** [comment]
Another issue with the `Send`/`Sync` bounds of *both*  guards: In the presence of upgrade/downgrade methods, the two guard types ***must*** have the same kind of (maximally restrictive) `Send` and `Sync` bounds on the `T` parameter, since each inherits the bounds from the other due to the fact that `upgrade` and `downgrade` exist.

**Comment by steffahn** [comment]
In conclusion: The first issue means _at least_ a change to
```rs
impl<T: Send, U: Send> Send for WriteGuard<'_, T, U>
impl<T, U: Sync> Sync for WriteGuard<'_, T, U>
```

whereas `ReadGuard` has
```rs
impl<T: Sync, U: Sync> Send for ReadGuard<'_, T, U>
impl<T: Sync, U: Sync> Sync for ReadGuard<'_, T, U>
```

---

The second issue then forces us to extend the `T: …` bounds to the union of both bounds, at least for the `Send` implementation of the guards, i.e.
```rs
impl<T: Send + Sync, U: Send> Send for WriteGuard<'_, T, U>
impl<T, U: Sync> Sync for WriteGuard<'_, T, U>

impl<T: Send + Sync, U: Sync> Send for ReadGuard<'_, T, U>
impl<T: Sync, U: Sync> Sync for ReadGuard<'_, T, U>
```

---

On another inspection of the API, I’m also noticing the `.rwlock` method on both guards. If you expose a `&TryRwLock<T>` which is only `Send` or `Sync` if `T: Send + Sync`, then better add those bounds _everywhere_ or else there would be a soundness whole, especially with the lifetime of the `&TryRwLock<T>` being the full `'locked` lifetime, so there is _no reasoning whatsoever_ about whether or not locks ensure any mutual exclusion, the `&'lock TryRwLock<T>` can stay valid long after the locks are dropped.

In conclusion, we need
```rs
impl<T: Send + Sync, U: Send> Send for WriteGuard<'_, T, U>
impl<T: Send + Sync, U: Sync> Sync for WriteGuard<'_, T, U>

impl<T: Send + Sync, U: Sync> Send for ReadGuard<'_, T, U>
impl<T: Send + Sync, U: Sync> Sync for ReadGuard<'_, T, U>
```

**Comment by SabrinaJewson** [comment]
Thank you very much for reporting this! It should be fixed in https://github.com/SabrinaJewson/try-rwlock.rs/commit/ccdec34ae26a93e0ab2b169e08f9ad2457f2890c.

**Comment by steffahn** [comment]
Looks good to me!
